src/array/cuda/labor_sampling.cu

```diff
@@ -0,0 +1,858 @@
+/*!
+ *   Copyright (c) 2022, NVIDIA Corporation
+ *   Copyright (c) 2022, GT-TDAlab (Muhammed Fatih Balin & Umit V. Catalyurek)  
+ *   All rights reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ *
+ * \file array/cuda/labor_sampling.cu
+ * \brief labor sampling
+ */
+
+#include <dgl/random.h>
+#include <dgl/runtime/device_api.h>
+#include <dgl/aten/coo.h>
+
+#include <thrust/execution_policy.h>
+#include <thrust/shuffle.h>
+#include <thrust/binary_search.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/copy.h>
+#include <thrust/transform.h>
+#include <thrust/zip_function.h>
+#include <thrust/gather.h>
+
+#include <curand_kernel.h>
+
+#include <numeric>
+#include <utility>
+#include <memory>
+#include <type_traits>
+#include <algorithm>
+#include <limits>
+
+#include "./dgl_cub.cuh"
+#include "./functor.cuh"
+#include "./spmm.cuh"
+
+#include "../../array/cuda/atomic.cuh"
+#include "../../runtime/cuda/cuda_common.h"
+#include "../../array/cuda/utils.h"
+#include "../../graph/transform/cuda/cuda_map_edges.cuh"
+
+namespace dgl {
+namespace aten {
+namespace impl {
+
+constexpr int BLOCK_SIZE = 128;
+constexpr int CTA_SIZE = 128;
+constexpr double eps = 0.0001;
+constexpr bool labor = true;
+
+namespace {
+
+template <typename ctx_t>
+class workspace_memory_alloc {
```

Comment:
Sure, any code that makes use of thrust can make use of this class. There are also helper allocation routines that use smart pointers so that lifetime of the allocated pointers are managed automatically.

Reply:
Where should I place the new header file?

Reply:
Sounds good.